packages/react/dialog/src/Dialog.tsx

```diff
@@ -164,19 +175,111 @@ const DialogContent = React.forwardRef((props, forwardedRef) => {
   const context = useDialogContext(CONTENT_NAME);
   return (
     <Presence present={forceMount || context.open}>
-      <DialogContentImpl data-state={getState(context.open)} {...contentProps} ref={forwardedRef} />
+      {context.modal ? (
+        <DialogContentModalImpl {...contentProps} ref={forwardedRef} />
+      ) : (
+        <DialogContentNonModalImpl {...contentProps} ref={forwardedRef} />
+      )}
     </Presence>
   );
 }) as DialogContentPrimitive;
 
+DialogContent.displayName = CONTENT_NAME;
+
+type DialogContentImplOwnProps = Omit<
+  Polymorphic.OwnProps<typeof DialogContentPart>,
+  'containFocus' | 'disableOutsidePointerEvents'
+>;
+
+type DialogContentImplPrimitive = Polymorphic.ForwardRefComponent<
+  Polymorphic.IntrinsicElement<typeof DialogContentPart>,
+  DialogContentImplOwnProps
+>;
+
+const DialogContentModalImpl = React.forwardRef((props, forwardedRef) => {
+  const context = useDialogContext(CONTENT_NAME);
+  const contentRef = React.useRef<HTMLDivElement>(null);
+  const composedRefs = useComposedRefs(forwardedRef, contentRef);
+
+  // aria-hide everything except the content (better supported equivalent to setting aria-modal)
+  React.useEffect(() => {
+    const content = contentRef.current;
+    if (content) return hideOthers(content);
+  }, []);
+
+  return (
+    <Portal>
```

Comment:
Annoyingly, we can't put the `Portal` as part of the `DialogContentPart` implementation because it needs to wrap the `RemoveScroll` component so that remove scroll works correctly…